org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextGeneratorTemplates.xtend

```diff
@@ -505,7 +505,7 @@ class XtextGeneratorTemplates {
 				
 				private static �activator.simpleName� INSTANCE;
 				
-				private �Map�<String, �Injector�> injectors = �Collections�.synchronizedMap(�Maps�.<String, �Injector�> newHashMapWithExpectedSize(1));
+				private �Map�<String, �Injector�> injectors = �Collections�.synchronizedMap(new �HashMap�<String, �Injector�>(2));
```

Comment:
Is there a lower for the Java version this template targets?
For example if a ConcurrentHashMap would be used here, below the method `getInjector(String)` could be simplified to:
```
public «Injector» getInjector(String language) {
	return injectors.computeIfAbsent(language, this::createInjector);
}
```
Which would also improve the concurrency since concurrent reads to not block each other and concurrent updates of **different keys** are possible.
But this requires at least Java-1.8.

Reply:
Xtext bundles require Java 11.

Reply:
> Xtext bundles require Java 11.

Yes, I wasn't sure if the projects whose code is generated with this template also depend on Xtext or maybe could be independent and therefore target even older java version. But I now have the impression that the former is the case.

Reply:
As usual, the devil is in the details. Injector creation may be re-entrant if a language tries to obtain services from another injector, e.g. from a super-language in the same bundle.

Hash collisions on strings are not super rare:

```java
public class ActivatorCodeTest {

	@Test
	public void testCollision() {
		String a = "com.acme.some.package.name.DB";
		String b = "com.acme.some.package.name.Ca";
		assertEquals(a.hashCode(), b.hashCode());
	}
	
	@Test
	public void testConcurrentMapComputeIfAbsent() {
		String a = "com.acme.some.package.name.DB";
		String b = "com.acme.some.package.name.Ca";
		
		Map<Object, Object> injectors = new ConcurrentHashMap<>();
		assertNotNull(injectors.computeIfAbsent(a, first->{
			return injectors.computeIfAbsent(b, reentrant->{
				return new Object();
			});
		}));
	}
	
	@Test
	public void testSynchronizedMapComputeIfAbsent() {
		String a = "com.acme.some.package.name.DB";
		String b = "com.acme.some.package.name.Ca";
		
		Map<Object, Object> injectors = Collections.synchronizedMap(new HashMap<>());
		assertNotNull(injectors.computeIfAbsent(a, first->{
			return injectors.computeIfAbsent(b, reentrant->{
				return new Object();
			});
		}));
	}
	
}
```

Reply:
What I try to say: Even if the code looks like a plain `computeIfAbsent`, it's not the same.

Reply:
> As usual, the devil is in the details. Injector creation may be re-entrant if a language tries to obtain services from another injector, e.g. from a super-language in the same bundle.

That would indeed be a problem.